### PR TITLE
[12.0][ADD] account_accountant_oca to allow to access to accouting settings

### PR DESCRIPTION
**Pull request by legalsylvain**
This module allows to set some accounting fields that are present in the model ``res.company`` but are not accessible like ``overdue_msg`` or only via the onboarding mechanism like ``fiscalyear_last_day``.

image

With this modules, that fields are available via Invoicing > Configuration > Settings > Invoicing

res_config_settings_fiscal_year_settings

The name of the module is ``account_accountant_oca`` because the feature are present in the module ``account_accountant`` that moved from Odoo 10.0 Community Edition to Odoo 11.0 Enterprise Edition.

### CODE REVIEW

**Review by remi-filament — APPROVED** [review]
Very useful to display these "hidden" fields !
Thank you Sylvain

**Review by HaraldPanten — APPROVED** [review]
LGTM 👍

**Review by rvalyi — APPROVED** [review]
code reviewed and tested

**Comment by pedrobaeza** [comment]
We already have `account_lock_date_update`: https://odoo-community.org/shop/product/account-lock-date-update-342

**Comment by legalsylvain** [comment]
Hi @pedrobaeza. 

> We already have account_lock_date_update: https://odoo-community.org/shop/product/account-lock-date-update-342

Well, 
- the objective of the module ``account_accountant_oca`` is to have the possibility to set all accounting settings that are unavailable in a fresh CE Instance. so not only ``period_lock_date`` and ``fiscalyear_lock_date`` but also ``fiscalyear_last_day``, ``fiscalyear_last_month``, ``overdue_msg``, ``incoterm_id`` and later all the other settings (see [Roadmap.rst](https://github.com/legalsylvain/account-financial-tools/blob/12.0-add-account_accountant_oca/account_accountant_oca/readme/ROADMAP.rst))
- the objectif of ``account_lock_date_update`` is only to handle ``period_lock_date`` and ``fiscalyear_lock_date``.
also in  ``account_lock_date_update`` it is via a popup and a wizard, that is not in the spririt of Odoo. 
In this module, it is in the regular place. (configuration > settings > accounting).

So my point of view is that ``account_accountant_oca`` is more generic. Also i think that for end users, it is more easy to install a single module that allow to set all the settings, than many little modules we have to search & install. 
I have similar point of view regarding missing accounting menus.
So, for the same reason, I prefer the V13 situation : (one single module ``account_menu`` that adds all the missing menus) than the V12 situation : 4 modules ``account_coa_menu``, ``account_group_menu``, ``account_tag_menu``, ``account_type_menu`` :  
-> Less work to maintain.
-> easier for users that are looking for such features.

**Comment by pedrobaeza** [comment]
OK, the problem with that approach is that only those with access to settings can change the lock date, and that's why the other module makes more sense, but both can coexist indeed.

**Comment by legalsylvain** [comment]
wait before merge please.

**Comment by legalsylvain** [comment]
> OK, the problem with that approach is that only those with access to settings can change the lock date, and that's why the other module makes more sense, but both can coexist indeed.

Well, indeed, I didn't thought about that. Thanks !
For me it's a general problem in Odoo. To configure odoo, in a business point of view, we have to be member of "Administration / Settings" group to access to ``res.config.settings`` part. But this model is mixing business things (see below) that should be available for "manager" and administrator things (mainly the possibility to install modules).

For accounting part, for example, It's not normal to have to be member of "Administration / settings" to configure taxes. Be member of "Accountant Manager" should be enough : 

image

For purchase part, a Purchase manager can not access to that configuration. He should : 
image
(``res_config_settings.lock_confirmed_po`` // ``res.company.po_lock``).

So, as this module (account_accountant_oca) takes the same approach as the Odoo one, it is reproducing the same problem. You're right.

In the other part, I keep thinking that having a single complete module to make all accounting settings is better than a lot of many little modules to maintain.

What about if I change this PR to : 
- add a new menu item "Invoicing > configuration > Accouting Settings" available for "Accountant Manager"
- add a new model ``account.config.settings`` that inherit from ``res.config.settings``.
- move the current view definition of this module, into a new view definition for this model ``account.config.settings``. 
In fact this model was existing until Odoo 10.0, but it disappeared after.

With such design : 
- there is a single module to all accounting configuration
- it is accessible for accounting manager

**Comment by pedrobaeza** [comment]
I see that the module becomes very high for simply things, so let's keep it as is, with its cons. We should speak with Odoo for improving that part instead. Not sure that all the settings require system settings group though in >=12.

**Comment by legalsylvain** [comment]
> I see that the module becomes very high for simply things, so let's keep it as is, with its cons. 

Not sure it will very more complicated. let me give a try, this end of week. 

> We should speak with Odoo for improving that part instead. 

You're right for sure. I don't know who to contact for such questions.

> Not sure that all the settings require system settings group though in >=12.

- V12 : ``base.group_system`` [Ref](https://github.com/odoo/odoo/blob/12.0/addons/account/views/res_config_settings_views.xml#L596). same for CRM, Sales, Purchase, Inventory and Point of Sale Settings menu at least.
- V13 : ``base.group_system``. [Ref](https://github.com/odoo/odoo/blob/13.0/addons/account/views/res_config_settings_views.xml#L539).
- master : ``base.group_system`` [Ref](https://github.com/odoo/odoo/blob/master/addons/account/views/account_menuitem.xml#L53).

File: account_accountant_oca/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+{
+    "name": "Account Accountant (OCA)",
+    "summary": "Add missing features for Odoo Community Edition"
+    " to configure properly accounting. (Fiscal years, Overdue Message, ...)",
+    "version": "12.0.1.0.1",
+    "category": "Accounting",
+    "author": "GRAP,Odoo Community Association (OCA)",
+    "maintainers": ["legalsylvain"],
+    "website": "http://www.github.com/OCA/account-financial-tools",
+    "license": "AGPL-3",
```
Comment (primes2h) [inline account_accountant_oca/__manifest__.py:14]: If this module is not compatible with `account_accountant`, you should add this line:

`    'excludes': ['account_accountant'],`

Comment (StefanRijnhart) [reply account_accountant_oca/__manifest__.py:14]: For most OCA developers it is not possible to know what is in which enterprise module. So I would prefer that we don't fill this list with enterprise modules as it is not possible to maintain it.

**Comment by StefanRijnhart** [comment]
@legalsylvain can we just merge this as is?

**Review by StefanRijnhart — APPROVED** [review]

**Review by marcelsavegnago — APPROVED** [review]

**Comment by legalsylvain** [comment]
Sorry I forgot this PR and I'm in hollydays. Please let me try what I said. I think it will cover more cases. I'll give a try 31 th August. If it's not ok, i'll reopen this PR to merge it.
Thanks.

**Comment by StefanRijnhart** [comment]
Thanks for your response, and enjoy your holiday!

**Comment by legalsylvain** [comment]
Hi,

My parental leaves are over and I'm back on that topic. 

I tried to make a refactoring following [my remark](https://github.com/OCA/account-financial-tools/pull/1004#issuecomment-638818793). specially to create a transient model ``_name = 'account.config.settings'``, ``_inherit = 'res.config.settings'`` 

it doesn't work because : 
- res.config.settings is a special object that are not available for non admin people
- the specific UI (data-key features) for this model is broken when renaming the model
- security is hard-coded in [python file](https://github.com/odoo/odoo/blob/12.0/odoo/addons/base/models/res_config.py#L615), and write a sudo could generate security issue.

For that purpose, I see the following options :

A) reopen this PR and merge it. I'm not confortable with that solution, because i'll not use that module ;-). the access problem is a blocking point I didn't saw but it makes the current design useless for my company. 

B) improve the current module ``account_lock_date_update``, adding extra accountant fields. (``fiscalyear_last_day``, ``fiscalyear_last_month``, ``overdue_msg``, ``incoterm_id``). it make the technical name obsolete, but it's not a big deal, and it could be updated in V14, when porting this module. One single module for all the accountant settings.

C) create a new module, with the same design as ``account_lock_date_update`` (a wizard for accountant people), and named ``account_accountant_setting``. Two concurent modules.(can be fixed in V14 also)

What do you think ? 

kind regards.

**Comment by legalsylvain** [comment]
Hi reviewers. Any idea regarding my last comment ? 

CC : @StefanRijnhart

**Comment by StefanRijnhart** [comment]
So Odoo combined all the config settings models into one and made it more difficult to create a custom settings page. But you might want to try to do it using Python object inheritance (misusing this model as as a mixin):

```
from odoo.addons.base.models.res_config import ResConfigSettings
class AccountingConfigSettings(ResConfigSettings):
    _name = 'accounting.config.settings'
```

If it works as I expect that will give you the core functionality of the model but without running the security risk of changing the settings in the original model. You'll have to overwrite the execute() method to apply the restrictions that are appropriate for your model.

Not sure what the UI breakage is but hopefully you don't run into it using this method.

**Comment by legalsylvain** [comment]
_Five years later ..._

Hi @StefanRijnhart, Happy new years !  I finally took the time to implement the solution you propose. There was some adaptation to do, but it finally works, so I wanted to thank you !

FYI, the PoC is here https://github.com/grap/grap-odoo-business/pull/177. (not an OCA candidate for the time being, because a lot of configuration are missing).

**Pull request closed without merging**